File: traffic/src/main/java/com/hsryuuu/traffic/tmp/coupon/entity/Coupon.java
package com.hsryuuu.traffic.tmp.coupon.entity;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Table(name = "coupon", uniqueConstraints = {
        @UniqueConstraint(columnNames = {"coupon_event_id", "user_id"})
})
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Coupon {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "coupon_event_id", nullable = false)
    private CouponEvent couponEvent;

    @Column(name = "user_id", nullable = false)
    private Long userId;

    @Column(nullable = false, updatable = false)
    private LocalDateTime issuedAt;

    private LocalDateTime usedAt;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private CouponStatus status;

    @Builder
    public Coupon(CouponEvent couponEvent, Long userId) {
        this.couponEvent = couponEvent;
        this.userId = userId;
        this.issuedAt = LocalDateTime.now();
        this.status = CouponStatus.ISSUED;
    }

    public void use() {
        this.status = CouponStatus.USED;
        this.usedAt = LocalDateTime.now();
    }

    public enum CouponStatus {
        ISSUED, USED, EXPIRED
    }
}

File: high-traffic/src/main/kotlin/com/hsryuuu/hightraffic/coupon/entity/Coupon.kt
package com.hsryuuu.hightraffic.coupon.entity

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table

@Entity
@Table(name = "coupon")
class Coupon(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @Column(nullable = false, length = 100)
    val name: String,

    @Column(name = "total_quantity", nullable = false)
    val totalQuantity: Int,

    @Column(name = "issued_quantity", nullable = false)
    var issuedQuantity: Int = 0,
) {
    val isSoldOut: Boolean
        get() = issuedQuantity >= totalQuantity

    fun issue() {
        issuedQuantity += 1
    }
}
